runtime/jailer.go

```diff
@@ -0,0 +1,476 @@
+	handler := j.BuildHandler(cfg, &machineConfig.SocketPath, vmID)
+	fifoHandler := j.BuildLinkFifoHandler()
+	client := firecracker.NewClient(machineConfig.SocketPath, j.logger, machineConfig.Debug)
+
+	opts := []firecracker.Opt{
+		firecracker.WithProcessRunner(j.jailerCommand(vmID)),
+		firecracker.WithClient(client),
+		func(m *firecracker.Machine) {
+			m.Handlers.FcInit = m.Handlers.FcInit.Prepend(handler)
+			m.Handlers.FcInit = m.Handlers.FcInit.AppendAfter(firecracker.CreateLogFilesHandlerName, fifoHandler)
+		},
+	}
+
+	return opts, nil
+}
+
+// BuildHandler will link the necessary files except for the fifos due to the
+// fifos needing to be created. Also, this will create the proper device nodes
+// needed by Firecracker
+func (j *runcJailer) BuildHandler(cfg *Config, socketPath *string, vmID string) firecracker.Handler {
+	ociBundlePath := j.OCIBundlePath()
+	rootPath := j.RootPath()
+	*socketPath = filepath.Join(rootPath, "api.socket")
+
+	return firecracker.Handler{
+		Name: jailerHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			const mode = os.FileMode(0700)
+			// Create the proper paths needed for the runc jailer
+			j.logger.WithField("rootPath", rootPath).Debug("Creating root drive path")
+			if err := os.MkdirAll(rootPath, mode); err != nil {
+				return errors.Wrapf(err, "failed to create root path: %v", rootPath)
+			}
+
+			binPath := filepath.Join(rootPath, "usr", "local", "bin")
+			j.logger.WithField("binPath", binPath).Debug("Creating /usr/local/bin")
+			if err := os.MkdirAll(binPath, mode); err != nil {
+				return errors.Wrapf(err, "failed to create /usr/local/bin in root path: %v", binPath)
+			}
+
+			contentsPath := j.ContentsPath()
+			j.logger.WithField("contentsPath", contentsPath).Debug("Creating firecracker contents path")
+			if err := os.MkdirAll(contentsPath, mode); err != nil {
+				return errors.Wrapf(err, "failed to create contents path: %v", contentsPath)
+			}
+
+			rootPathToConfig := filepath.Join(ociBundlePath, "config.json")
+			j.logger.WithField("rootPathToConfig", rootPathToConfig).Debug("Copying config")
+			if err := copyFile(runcConfigPath, rootPathToConfig, 0666); err != nil {
+				return errors.Wrapf(err, "failed to copy config from %v to %v", runcConfigPath, rootPathToConfig)
+			}
+
+			j.logger.Debug("Overwritting process args of config")
+			if err := j.overwriteConfig(cfg, filepath.Base(m.Cfg.SocketPath), rootPathToConfig); err != nil {
+				return errors.Wrap(err, "failed to overwrite config.json")
+			}
+
+			// copy the firecracker binary
+			j.logger.WithField("binPath", binPath).Debug("copying firecracker binary")
+			if err := copyFile(cfg.FirecrackerBinaryPath, binPath, 0700); err != nil {
+				return errors.Wrapf(err, "could not copy firecracker binary from path %v", cfg.FirecrackerBinaryPath)
+			}
+
+			// Link the kernel image
```

Comment:
Why do we link the kernel img but copy the firecracker binary? It seems like all the security tradeoffs between copying vs. linking/bind-mounting are the same between the two, so I don't understand why we would end up using different approaches for the firecracker binary vs. the kernel img.

In the case that linking is preferable here vs copying, I'd again prefer we do a bind-mount instead of linking. bind-mounts don't have the same limitation of being between the same filesystem that hardlinks do and bind-mounts allow you to mount as read-only, which gives an extra layer of protection

Reply:
This logic is from the SDK's jailer stuff. I can't seem to get `syscall.Mount` to work properly, so I'd rather leave it using `Link` instead for now. We are going to do another pass at this to include user namespaces and making things more secure. I think that'll be a good time to include also bind mounting rather than `Link` files.

Reply:
So we are just going to replicate Firecracker's jailer for now and then make updates in a second pass? That's sounds good to me, I'm still confused why they would mix copying and linking but as long as it's addressed in a next pass that's fine for now.

Reply:
+1 to @sipsma's comments about using a bind mount instead of a hardlink. @xibz, are you tracking the to-do list for revision 2 anywhere?